(CFACT-169) Add install task for cfacter.rb

MikaelSmith opened the pull request:
Install cfacter into the current Ruby if it's available. Added to
support installing a cfacter package that includes Ruby functionality.

melissaanne commented:
What about the `FindRuby` module you found?

melissaanne commented:
Oh wait, nope, my bad, this is it :)

melissaanne commented:
:+1: it looks good to me! I'm going to test this how this performs on linux, and I'll let you know the results

melissaanne commented:
Huh, I'm getting 
```
-- The following OPTIONAL packages have not been found:

 * Ruby (required version >= 1.9)
```
even though we have `ruby.x86_64 0:2.0.0.353-22.el7_` installed. I want to look into this further...

MikaelSmith commented:
The CMake documentation is at http://www.cmake.org/cmake/help/v3.0/module/FindRuby.html; it's not particularly extensive, so I can spend some time looking tomorrow.

melissaanne commented:
I'm still trying to figure this out, and I'm not getting it to work. Just confirming, you got this to work successfully @MikaelSmith ? What version of ruby was on the system you were building on?

MikaelSmith commented:
I was using it on Mac, rbenv, Ruby 2.1.5.

melissaanne commented:
I just ran this manually (aka not in a mock/cow environment), and it was successful. I'm going to continue poking around to see if I can figure out why this isn't working for our cows/mocks, and get back to you

melissaanne commented:
Also, it looks like we'll need to upgrade cmake to 3.0.0 if we want the findruby module to work for ruby 2

MikaelSmith commented:
Will that be difficult? We could add a copy of the updated FindRuby.cmake to our repository.

melissaanne commented:
No, it shouldn't be too difficult. I just means shipping a new release, which will take me a few days. But that doesn't explain why it's not working when ruby 1.9 is available in the mock/cow.

melissaanne commented:
:+1:  fully now, I got this working. We're not going to see any changes until I get my pull request in to update the build time requirements we're pulling in, but we will not see any failures in the mean time.

MikaelSmith commented on lib/tests/CMakeLists.txt:
```diff
@@ -44,7 +44,7 @@ set(LIBFACTER_TESTS_COMMON_SOURCES
 
 # Add the ruby tests if there's a ruby installed
-if (RUBY_VERSION VERSION_GREATER '1.8.7')
+if (RUBY_FOUND)
```
@peterhuene What reason did you use the previous pattern? Support for Ruby 2+ in cmake 2.8.x? Or another reason?

peterhuene replied:
To prevent ruby tests from running on a system with 1.8?  Is there a particular problem with this approach? I haven't been following this issue.

MikaelSmith replied:
Only that RUBY_FOUND isn't true if the library or include_dir isn't found. But the FindRuby module takes care of checking the version number as well, and seems like the more obvious way.

peterhuene replied:
Ah, I see.  Yeah, if it sets `RUBY_EXECUTABLE` without the includes, we should be using that.

MikaelSmith commented on CMakeLists.txt:
```diff
@@ -46,6 +46,14 @@ if (WIN32)
 find_package(Boost 1.54 REQUIRED COMPONENTS ${BOOST_PKGS})
 
+find_package(Ruby 1.9)
+if (RUBY_FOUND)
```
I'll change this to just verify a RUBY_EXECUTABLE of the right version is found.

melissaanne replied:
As long as we're installing ruby and ruby-devel, `RUBY_FOUND` should be returning as true

MikaelSmith commented:
It seems fair to expect libruby dynamic library to exist if you're developing, so going to leave these the way they are.

MikaelSmith commented:
I'm wary of this because it makes installing to Ruby implicit, and isn't obvious how to install to a different Ruby if you want to test with multiple Ruby versions.

MikaelSmith commented:
@melissaanne addressed those concerns:
* install writes anything anyway, adding a library to Ruby's vendor dir isn't surprising
* installing to a different Ruby requires repeating cmake configuration, I'll document that

haus commented:
:+1:
